src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -1215,59 +1215,84 @@ public override BoundNode VisitNullCoalescingOperator(BoundNullCoalescingOperato
         {
             Debug.Assert(!IsConditionalState);
 
-            VisitRvalue(node.LeftOperand);
-            if (IsConstantNull(node.LeftOperand))
+            BoundExpression leftOperand = node.LeftOperand;
+            BoundExpression rightOperand = node.RightOperand;
+
+            Result leftResult = VisitRvalueWithResult(leftOperand);
+            if (IsConstantNull(leftOperand))
             {
-                VisitRvalue(node.RightOperand);
+                VisitRvalue(rightOperand);
                 return null;
             }
 
             var leftState = this.State.Clone();
-            var leftType = _result.Type;
-            if (leftType?.IsNullable == false)
+            if (leftResult.Type?.IsNullable == false)
             {
-                ReportStaticNullCheckingDiagnostics(ErrorCode.HDN_ExpressionIsProbablyNeverNull, node.LeftOperand.Syntax);
+                ReportStaticNullCheckingDiagnostics(ErrorCode.HDN_ExpressionIsProbablyNeverNull, leftOperand.Syntax);
             }
 
-            bool leftIsConstant = node.LeftOperand.ConstantValue != null;
+            bool leftIsConstant = leftOperand.ConstantValue != null;
             if (leftIsConstant)
             {
                 SetUnreachable();
             }
 
-            leftType = InferResultNullability(node.LeftOperand, node.LeftConversion, node.Type, leftType);
-
-            VisitRvalue(node.RightOperand);
-            var rightType = _result.Type;
-
+            var rightResult = VisitRvalueWithResult(rightOperand);
             IntersectWith(ref this.State, ref leftState);
 
+            bool? resultIsNullable = getIsNullable(leftOperand, leftResult) == false ? false : getIsNullable(rightOperand, rightResult);
             TypeSymbolWithAnnotations resultType;
-
-            if (node.Type.IsErrorType())
+            switch (node.OperatorResultKind)
             {
-                resultType = TypeSymbolWithAnnotations.Create(node.Type, isNullableIfReferenceType: null);
+                case BoundNullCoalescingOperatorResultKind.NoCommonType:
+                    resultType = TypeSymbolWithAnnotations.Create(node.Type);
+                    break;
+                case BoundNullCoalescingOperatorResultKind.LeftType:
+                    resultType = TypeSymbolWithAnnotations.Create(getLeftResultType(leftResult.Type.TypeSymbol, rightResult.Type?.TypeSymbol), resultIsNullable);
+                    break;
+                case BoundNullCoalescingOperatorResultKind.LeftUnwrappedType:
+                    resultType = TypeSymbolWithAnnotations.Create(getLeftResultType(unwrapIfNullable(leftResult.Type.TypeSymbol), rightResult.Type?.TypeSymbol), resultIsNullable);
+                    break;
+                case BoundNullCoalescingOperatorResultKind.RightType:
+                    resultType = TypeSymbolWithAnnotations.Create(getRightResultType(leftResult.Type.TypeSymbol, rightResult.Type.TypeSymbol));
```

Comment:
>TypeSymbolWithAnnotations.Create(getRightResultType(leftResult.Type.TypeSymbol, rightResult.Type.TypeSymbol)); [](start = 33, length = 110)

Are we losing resultIsNullable here?

Reply:
Yes, thanks. Fixed.

---
In reply to: [175210162](https://github.com/dotnet/roslyn/pull/25358#discussion_r175210162) [](ancestors = 175210162)